sklearn/linear_model/cd_fast.pyx

```diff
@@ -175,108 +199,212 @@ def enet_coordinate_descent(np.ndarray[DOUBLE, ndim=1] w,
-
-                if fabs(w[ii]) > w_max:
-                    w_max = fabs(w[ii])
-
-            if (w_max == 0.0
-                    or d_w_max / w_max < d_w_tol
-                    or n_iter == max_iter - 1):
-                # the biggest coordinate update of this iteration was smaller
-                # than the tolerance: check the duality gap as ultimate
-                # stopping criterion
-
-                # XtA = np.dot(X.T, R) - beta * w
-                for i in range(n_features):
-                    XtA[i] = ddot(
-                        n_samples,
-                        <DOUBLE*>(X.data + i * n_samples *sizeof(DOUBLE)),
-                        1, <DOUBLE*>R.data, 1) - beta * w[i]
-
-                if positive:
-                    dual_norm_XtA = max(n_features, <DOUBLE*>XtA.data)
-                else:
-                    dual_norm_XtA = abs_max(n_features, <DOUBLE*>XtA.data)
-
-                # R_norm2 = np.dot(R, R)
-                R_norm2 = ddot(n_samples, <DOUBLE*>R.data, 1,
-                               <DOUBLE*>R.data, 1)
-
-                # w_norm2 = np.dot(w, w)
-                w_norm2 = ddot(n_features, <DOUBLE*>w.data, 1,
-                               <DOUBLE*>w.data, 1)
-
-                if (dual_norm_XtA > alpha):
-                    const = alpha / dual_norm_XtA
-                    A_norm2 = R_norm2 * (const ** 2)
-                    gap = 0.5 * (R_norm2 + A_norm2)
-                else:
-                    const = 1.0
-                    gap = R_norm2
-
-                l1_norm = dasum(n_features, <DOUBLE*>w.data, 1)
-
-                # np.dot(R.T, y)
-                gap += (alpha * l1_norm - const * ddot(
+        if floating is double:
+            # R = y - np.dot(X, w)
+            for i in range(n_samples):
+                R[i] = y[i] - ddot(n_features,
+                                <DOUBLE*>(X.data + i * sizeof(DOUBLE)),
+                                n_samples, <DOUBLE*>w.data, 1)
+
+            # tol *= np.dot(y, y)
+            tol *= ddot(n_samples, <DOUBLE*>y.data, n_tasks,
+                        <DOUBLE*>y.data, n_tasks)
+
+            for n_iter in range(max_iter):
+                w_max = 0.0
+                d_w_max = 0.0
+                for f_iter in range(n_features):  # Loop over coordinates
+                    if random:
+                        ii = rand_int(n_features, rand_r_state)
+                    else:
+                        ii = f_iter
+
+                    if norm_cols_X[ii] == 0.0:
+                        continue
+
+                    w_ii = w[ii]  # Store previous value
+
+                    if w_ii != 0.0:
+                        # R += w_ii * X[:,ii]
+                        daxpy(n_samples, w_ii,
+                            <DOUBLE*>(X.data + ii * n_samples * sizeof(DOUBLE)),
+                            1, <DOUBLE*>R.data, 1)
+
+                    # tmp = (X[:,ii]*R).sum()
+                    tmp = ddot(n_samples,
```

Comment:
If these BLAS functions are the reason you use the if/else case, you could locally do `ddot if floating is double else sdot` or perhaps make a flag variable for if `floating is double` to make this briefer. Indeed it might be possible to declare a function pointer with `floating *` parameters above, set it to `dot = ddot if floating is double else sdot` and see if the C compiler optimises out the fact that the pointer is never changed.